ddtrace/contrib/urllib3/patch.py

```diff
@@ -0,0 +1,131 @@
+import urllib3
+
+from ddtrace import config
+from ddtrace.pin import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from .. import trace_utils
+from ...compat import parse
+from ...constants import ANALYTICS_SAMPLE_RATE_KEY
+from ...ext import SpanTypes
+from ...ext import http
+from ...propagation.http import HTTPPropagator
+from ...utils import ArgumentError
+from ...utils import get_argument_value
+from ...utils.formats import asbool
+from ...utils.formats import get_env
+from ...utils.wrappers import unwrap as _u
+
+
+# Ports which, if set, will not be used in hostnames/service names
+DROP_PORTS = (80, 443)
+
+# Initialize the default config vars
+config._add(
+    "urllib3",
+    {
+        "_default_service": "urllib3",
+        "distributed_tracing": asbool(get_env("urllib3", "distributed_tracing", default=True)),
+        "analytics_enabled": asbool(get_env("urllib3", "analytics_enabled", default=False)),
+        "analytics_sample_rate": float(get_env("urllib3", "analytics_sample_rate", default=1.0)),
+        "trace_query_string": asbool(get_env("urllib3", "trace_query_string", default=False)),
+        "split_by_domain": asbool(get_env("urllib3", "split_by_domain", default=False)),
+    },
+)
+
+
+def patch():
+    """Enable tracing for all urllib3 requests"""
+    if getattr(urllib3, "__datadog_patch", False):
+        return
+    setattr(urllib3, "__datadog_patch", True)
+
+    _w("urllib3", "connectionpool.HTTPConnectionPool.urlopen", _wrap_urlopen)
+
+
+def unpatch():
+    """Disable trace for all urllib3 requests"""
+    if getattr(urllib3, "__datadog_patch", False):
+        setattr(urllib3, "__datadog_patch", False)
+
+        _u(urllib3.connectionpool.HTTPConnectionPool, "urlopen")
+
+
+def _wrap_urlopen(func, instance, args, kwargs):
+    """
+    Wrapper function for the lower-level urlopen in urllib3
+
+    :param func: The original target function "urlopen"
+    :param instance: The patched instance of ``HTTPConnectionPool``
+    :param args: Positional arguments from the target function
+    :param kwargs: Keyword arguments from the target function
+    :return: The ``HTTPResponse`` from the target function
+    """
+    request_method = get_argument_value(args, kwargs, 0, "method")
+    request_url = get_argument_value(args, kwargs, 1, "url")
+    try:
+        request_headers = get_argument_value(args, kwargs, 3, "headers")
+    except ArgumentError:
+        request_headers = None
+    try:
+        request_retries = get_argument_value(args, kwargs, 4, "retries")
+    except ArgumentError:
+        request_retries = None
+
+    # HTTPConnectionPool allows relative path requests; convert the request_url to an absolute url
+    if request_url.startswith("/"):
+        request_url = parse.urlunparse(
+            (
+                instance.scheme,
```

Comment:
We should be defensive here and catch any `TypeError` which had been an issue with similar code in the django integration.

https://github.com/python/cpython/blob/a2f0654b0a5b4c4f726155620002cc1f5f2d206a/Lib/urllib/parse.py#L122

Reply:
Looking at https://github.com/urllib3/urllib3/blob/44ee4ad281444ec0b0a0aeeb2c8bf423c15cbc20/src/urllib3/connectionpool.py#L1004, it seems that all the URL pieces are either `None` or instances of `str`, so we should be good here.